### DISCUSSION TITLE
Autocompletion for __init_subclass__ (PEP 487)

### DISCUSSION THREAD

**Discussion by timrid** [microsoft/pylance-release#2781] [Enhancement]
I have seen in microsoft/pyright#1020 that PEP487 is supported. This also works for me, so that the types are recognized correctly and also if parameters are missing this is recognized. But what I still miss is that missing parameters are not suggested by the autocompletion. Is it possible to add this function?

With the following example, I would expect that the autocompletion suggests `bar=`:
```python
class Foo:
    def __init_subclass__(cls, *, bar: str) -> None:
        super().__init_subclass__()
        # Do something with bar

class Bar(Foo, b<---Hit Ctrl+Space here
```

**Comment by erictraut** [comment]
I think this is a relatively rare use case, so I don't know if it makes sense to add completion suggestions specifically for this case, but we would consider it if there's sufficient user interest.

Since this feature request isn't related to core type checking, I'm going to transfer it to the pylance-release project. We generally prioritize features like this one based on user feedback (upvotes).

**Comment by erictraut** [comment]
@judej, you may want to move this to a discussion topic to see if there is interest from other pylance users. I don't think it needs any investigation.

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.

**Comment by NickCrews** [comment]
I'm more interested in the use case of getting completions on the resulting class:

```python

class Foo:
    def __init_subclass__(cls **kwargs: Any) -> None:
        super().__init_subclass__(**kwargs)
        cls.custom_attribute = 9

class Bar(Foo):
    def __init__(self):
        baz = self.custo<I would love autocompletion/type hints here>
```

**Reply by erictraut** [reply]
@NickCrews, this is a different case. The functionality requested by the OP requires custom handling in the completion provider (the component that provides completion suggestions). Your case is explained by a bug in the type analyzer logic within pyright, the type checker that pylance is built upon. Pyright is not treating `__init_subclass__` as a class method, but it should. This method, along with `__class_getitem__`, needs to be special-cased because it's implicitly a class method even though it doesn't use a `@classmethod` decorator. I've updated pyright to handle this case. This will be included in the next release of pyright and a future release of pylance.

**Reply by NickCrews** [reply]
Wow, thanks for the really fast fix @erictraut !
